File: tks/RepositoriesAdapters/src/main/kotlin/repository/adapters/RentRepositoryAdapter.java
package repository.adapters;

import repository.data.AbstractResourceEntity;
import repository.data.RentEntity;
import repository.data.UserEntity;
import ports.secondary.RentPersistencePort;
import ports.secondary.RentSearchPort;
import repository.mappers.RentMapper;
import domain.model.Rent;
import domain.model.values.AccessionNumber;
import repository.repositories.RepositoryBase;

import java.util.Objects;

public class RentRepositoryAdapter implements RentSearchPort, RentPersistencePort {

    private final RepositoryBase<RentEntity> repository;
    private final RepositoryBase<AbstractResourceEntity> resourceRepository;
    private final RepositoryBase<UserEntity> userRepository;
    private final RentMapper mapper;

    public RentRepositoryAdapter(RepositoryBase<RentEntity> repository, RepositoryBase<AbstractResourceEntity> resourceRepository, RepositoryBase<UserEntity> userRepository, RentMapper mapper) {
        this.repository = repository;
        this.resourceRepository = resourceRepository;
        this.userRepository = userRepository;
        this.mapper = mapper;
    }

    @Override
    public void save(Rent rent) {
        var entity = repository.find(x -> Objects.equals(x.getId(), rent.getId()));
        if(entity == null){
            entity = mapper.mapDomainObjectToEntity(rent);
            var resource = resourceRepository.find(x -> Objects.equals(x.getAccessionNumber(), rent.getResourceId()));
            var emailValue = rent.getUserEmail().getValue();
            var user = userRepository.find(x -> x.getEmail().equals(emailValue));
            // todo should nullability be checked ?
            entity.setResource(resource);
            entity.setUser(user);
            repository.add(entity);
        }else{
            mapper.mapDomainObjectToEntity(rent, entity);
            repository.update(entity);
        }
    }

    @Override
    public Rent findActiveByResourceId(AccessionNumber resourceId) {
        var accessionNumberValue = resourceId.getValue();
        var entity = repository.find(x -> Objects.equals(x.getResource().getAccessionNumber(), accessionNumberValue) && x.getEndDate() == null);
        return mapper.mapEntityToDomainObject(entity);
    }
}

File: tks/RepositoriesAdapters/src/main/kotlin/repository/adapters/RentRepositoryAdapter.kt
package repository.adapters

import domain.model.context.rents.Rent
import domain.model.values.AccessionNumber
import ports.secondary.IRentRepositoryAdapter
import repository.data.RentEntity
import repository.data.ClientEntity
import repository.data.ProductEntity
import repository.mappers.RentMapper
import repository.repositories.IRepository
import java.util.*

class RentRepositoryAdapter(
    private val repository: IRepository<RentEntity>,
    private val resourceRepository: IRepository<ProductEntity>,
    private val userRepository: IRepository<ClientEntity>,
    private val mapper: RentMapper
) : IRentRepositoryAdapter {

    override fun save(rent: Rent) {
        var entity = repository.find { x: RentEntity -> x.id == rent.id }
        if (entity == null) {
            entity = mapper.mapDomainObjectToEntity(rent)!!
            val resource = resourceRepository.find { x: ProductEntity -> x.accessionNumber == rent.resourceId.value }
            val user = userRepository.find { (_, email) -> email == rent.userEmail.value }
            entity.resource = resource
            entity.user = user
            repository.add(entity)
        } else {
            // user and resource are ignored explicitly
            mapper.mapDomainObjectToEntity(rent, entity)
            repository.update(entity)
        }
    }

    override fun findActiveByResourceId(accessionNumber: AccessionNumber): Rent? {
        val entity =
            repository.find { x: RentEntity -> x.resource?.accessionNumber == accessionNumber.value && x.endDate == null }
        return mapper.mapEntityToDomainObject(entity)
    }

    override fun getById(id: UUID): Rent? {
        val entity = repository.find { x: RentEntity -> x.id == id }
        return mapper.mapEntityToDomainObject(entity)
    }
}